### PR TITLE
Add ExternalConsentManagement to application REST API

### PR DESCRIPTION
**Pull request by sandushi**
## Purpose
- Externalized consent management configuration is added to the application REST API
- Payload is like below.
```json
 "externalConsentManagement": {
            "enabled": true,
            "consentPageUrl": "https://example.com/consent"
        }
```
- To add `useExternalConsentManagement` and `externalConsentURL` properties to the SP meata data table from  application rest API

## Goals
- Add new property called useExternalConsentManagement  and externalConsentURL to application PATCH. POST, GET APIs

## Approach
- Add the property under AdvancedConfigurationProperty in PATCH. POST, GET calls

## Related PRs
- **This should merge after https://github.com/wso2/carbon-identity-framework/pull/4475**

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by mefarazath** [comment]
@sandushi can you add a sample POST and GET payload to show how the payloads will change with your PR?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 1]
```diff
@@ -2578,6 +2578,14 @@ components:
           type: boolean
           example: false
           description: Decides whether user consent needs to be skipped during logout flows.
+        useExternalConsentManagement:
+          type: boolean
+          example: false
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2583]: Can we set a default value to false

Comment (sandushi) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2583]: I have set the default value as false for this.

Comment (mefarazath) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2583]: we can set a value in swagger as well right?

Comment (sandushi) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: Resolved by https://github.com/wso2/identity-api-server/pull/434/commits/eee60c5adeaf9335a30d63b7f64010aa3d601cf6

**Comment by sandushi** [comment]
### GET Request payload after these changes

```json
{
    "id": "0c9bc30b-fbd0-4835-8339-2e071666f1af",
    "name": "oidc-app-3",
    "clientId": "IGhFOwdAOz8DXy70wnh7AUmJR3ca",
    "issuer": "",
    "templateId": "b9c5e11e-fc78-484b-9bec-015d247561b8",
    "isManagementApp": false,
    "claimConfiguration": {
        "dialect": "LOCAL",
        "claimMappings": [
            {
                "applicationClaim": "http://wso2.org/claims/username",
                "localClaim": {
                    "uri": "http://wso2.org/claims/username"
                }
            }
        ],
        "requestedClaims": [
            {
                "claim": {
                    "uri": "http://wso2.org/claims/username"
                },
                "mandatory": false
            }
        ],
        "subject": {
            "claim": {
                "uri": "http://wso2.org/claims/username"
            },
            "includeUserDomain": false,
            "includeTenantDomain": false,
            "useMappedLocalSubject": false
        },
        "role": {
            "includeUserDomain": true,
            "claim": {
                "uri": "http://wso2.org/claims/role"
            }
        }
    },
    "inboundProtocols": [
        {
            "type": "oauth2",
            "self": "/t/sandu12/api/server/v1/applications/0c9bc30b-fbd0-4835-8339-2e071666f1af/inbound-protocols/oidc"
        }
    ],
    "authenticationSequence": {
        "type": "DEFAULT",
        "steps": [
            {
                "id": 1,
                "options": [
                    {
                        "idp": "LOCAL",
                        "authenticator": "BasicAuthenticator"
                    }
                ]
            }
        ],
        "requestPathAuthenticators": [],
        "subjectStepId": 1,
        "attributeStepId": 1
    },
    "advancedConfigurations": {
        "saas": false,
        "discoverableByEndUsers": false,
        "skipLoginConsent": false,
        "skipLogoutConsent": false,
        "returnAuthenticatedIdpList": false,
        "enableAuthorization": false,
        "fragment": false,
        "additionalSpProperties": [],
        "externalConsentManagement": {
            "enabled": true,
            "consentUrl": "https://example.com/consent"
        }
    },
    "provisioningConfigurations": {
        "outboundProvisioningIdps": []
    },
    "access": "WRITE"
}
```

### POST Request Payload after the Changes

```json
{
    "name": "sample-oidc-app",
    "advancedConfigurations": {
        "discoverableByEndUsers": false,
        "skipLogoutConsent": true,
        "externalConsentManagement": {
            "enabled": true,
            "consentUrl": "https://example.com/consent"
        }
    },
    "authenticationSequence": {
        "type": "DEFAULT",
        "steps": [
            {
                "id": 1,
                "options": [
                    {
                        "idp": "LOCAL",
                        "authenticator": "basic"
                    }
                ]
            }
        ]
    },
    "claimConfiguration": {
        "dialect": "LOCAL",
        "requestedClaims": [
            {
                "claim": {
                    "uri": "http://wso2.org/claims/username"
                }
            }
        ]
    },
    "inboundProtocolConfiguration": {
        "oidc": {
            "grantTypes": [
                "authorization_code"
            ],
            "allowedOrigins": [
                "http://localhost:8080"
            ],
            "callbackURLs": [
                "http://localhost:8080/oidc-sample-app-1/oauth2client"
            ],
            "publicClient": false
        }
    },
    "templateId": "b9c5e11e-fc78-484b-9bec-015d247561b8"
}
```

### PATCH Request to update the ExternalConsentManagement configuration after the changes

**Review by mefarazath — DISMISSED** [review]

**Review by Wathsara — DISMISSED** [review]

**Review by mefarazath — APPROVED** [review]

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 2]
```diff
@@ -2711,6 +2713,19 @@ components:
...
+    ExternalizedConsentPage:
+      type: object
+      properties:
+        enabled:
+          type: boolean
+          description: >-
+            Decides whether external consent management is enabled.
+          default: false
+        consentPageUrl:
+          type: string
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2725]: is there a way to validate this to be a valid URL?

Comment (sandushi) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2637]: Resolved from https://github.com/wso2/identity-api-server/pull/434/commits/0180146d25d577568965efdc48757dd43680f63e

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java [hunk 3]
```diff
@@ -79,6 +84,51 @@ private void updateCertificate(Certificate certificate, ServiceProvider serviceP
...
+
+        ExternalizedConsentPageConfig externalConsentManagementConfig = getExternalizedConsentPageConfig(config);
+        if (externalizedConsentPageApiModel != null) {
+            if (isValidConsentPageUrl(externalizedConsentPageApiModel.getConsentPageUrl())) {
+                setIfNotNull(externalizedConsentPageApiModel.getEnabled(),
+                        externalConsentManagementConfig::setEnabled);
+                setIfNotNull(externalizedConsentPageApiModel.getConsentPageUrl(),
+                        externalConsentManagementConfig::setConsentPageUrl);
+            } else {
+                throw buildBadRequestError("Only https consent page urls are allowed ");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java:98]: Is the error message format consistent. (There is a space at the end)

**Review by mefarazath — APPROVED** [review]

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java [hunk 4]
```diff
@@ -79,6 +84,51 @@ private void updateCertificate(Certificate certificate, ServiceProvider serviceP
...
+
+        try {
+            URL url = new URL(consentPageUrl);
+            if ("https".equals(url.getProtocol())) {
+                return true;
+            } else {
+                return false;
+            }
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("Invalid consent page URL.");
```
[the remark this answers is not in the document]
Comment (sandushi) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java]: Resolved by https://github.com/wso2/identity-api-server/pull/434/commits/98196e6ff5d4749f1af21cc83950974389c04625

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java (hunk 3 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAdvancedConfigurations.java]: Resolved by https://github.com/wso2/identity-api-server/pull/434/commits/98196e6ff5d4749f1af21cc83950974389c04625

**Review by Wathsara — APPROVED** [review]

**Pull request merged**